From 034a79b3f9e48b7c534bf9ed0146524b379ff9fb Mon Sep 17 00:00:00 2001
From: Sebastian Birunt <sebastian.birunt@arm.com>
Date: Tue, 16 May 2023 13:30:54 +0200
Subject: [PATCH] feat: visit modules with position

Currntly Blueprint does not provide (exports) any
information about the module position from `build.bp`
file where it's defined.

However some external utilities want to have modules
position for a better management.

Provide ability to walk all modules with its position.
---
 context.go | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/context.go b/context.go
index da86837..ee0cf34 100644
--- a/context.go
+++ b/context.go
@@ -2751,6 +2751,23 @@ func (c *Context) visitAllModules(visit func(Module)) {
 	}
 }
 
+func (c *Context) visitAllModulesWithPos(visit func(Module, scanner.Position)) {
+	var module *moduleInfo
+
+	defer func() {
+		if r := recover(); r != nil {
+			panic(newPanicErrorf(r, "VisitAllModules(%s) for %s",
+				funcName(visit), module))
+		}
+	}()
+
+	for _, moduleGroup := range c.sortedModuleGroups() {
+		for _, module = range moduleGroup.modules {
+			visit(module.logicModule, module.pos)
+		}
+	}
+}
+
 func (c *Context) visitAllModulesIf(pred func(Module) bool,
 	visit func(Module)) {
 
@@ -3051,6 +3068,10 @@ func (c *Context) VisitAllModules(visit func(Module)) {
 	c.visitAllModules(visit)
 }
 
+func (c *Context) VisitAllModulesWithPos(visit func(Module, scanner.Position)) {
+	c.visitAllModulesWithPos(visit)
+}
+
 func (c *Context) VisitAllModulesIf(pred func(Module) bool,
 	visit func(Module)) {
 
-- 
2.25.1

